### ISSUE TITLE
Feedback problem 

### ISSUE THREAD

**Issue by TCQQ** [Gericop/Android-Support-Preference-V7-Fix#92]
material design specification the dividing line should be added below the grouping，how to add the division line below the group?

Like this：
https://material.io/guidelines/patterns/settings.html#settings-usage

**Comment by Gericop** [comment]
Use the `DIVIDER_CATEGORY_BEFORE_FIRST | DIVIDER_CATEGORY_BETWEEN` flags to achieve the desired behavior.

A complete example for your fragment extending `PreferenceFragmentCompatDividers`:

```java
@Override
public View onCreateView(LayoutInflater inflater, ViewGroup container, @Nullable Bundle savedInstanceState) {
    try {
        return super.onCreateView(inflater, container, savedInstanceState);
    } finally {
        setDividerPreferences(DIVIDER_CATEGORY_BEFORE_FIRST | DIVIDER_CATEGORY_BETWEEN);
    }
}
```

As for the independent lines between certain "groups" without the group titles (e.g. in your example the line between the "Do not disturb" and "Phone ringtone" preferences), it's not possible because if you don't specify the `title` for the `PreferenceCategory`, it will still use the same layout without the title being there, which makes it look like a huge empty space. Is this something you want to also include or you just wanted to know the config for the divider flags?

**Comment by TCQQ** [comment]
Ok thanks

**Comment by TCQQ** [comment]
But when I check the CheckBox, then I turn the phone into a horizontal screen, CheckBox status will be changed to not checked

**Comment by Gericop** [comment]
You probably set the `persistent` attribute to `false` for the `CheckBoxPreference` in your XML.

**Comment by TCQQ** [comment]
No, the Activity will be reset when the phone screen rotates

**Comment by TCQQ** [comment]
When the Activity is reset, the state of the control will not be saved

**Comment by TCQQ** [comment]
Such as the screen before the rotation of the control state is true, and then I click to false, and then when the screen rotation, the control state will become true

**Comment by Gericop** [comment]
If you have the `persistent` attribute set to `true` (default), the value will be saved immediately. Note, however, that you need to specify the `key` for the preference in order to be saved.

**Comment by TCQQ** [comment]
i dont understand

**Comment by Gericop** [comment]
Please, post your preferences XML. You probably didn't setup the XML well.

**Comment by TCQQ** [comment]
I set persistent to true  , and when the Activity is reset, the control state will not be saved

**Comment by TCQQ** [comment]
<?xml version="1.0" encoding="utf-8"?>
<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto">
    <PreferenceCategory android:title="First category">
        <Preference
            android:persistent="true"
            android:summary="This is just a normal preference"
            android:title="Preference" />
        <Preference
            android:enabled="false"
            android:persistent="true"
            android:summary="This is just a disabled preference"
            android:title="Preference 2" />
        <Preference
            android:icon="@drawable/ic_battery_std_black_24dp"
            android:persistent="true"
            android:summary="This is a preference with icon"
            android:title="Preference 3" />
        <Preference
            android:persistent="true"
            android:summary="This is just a preference with its icon space reserved"
            android:title="Preference 4"
            app:iconSpaceReserved="true" />
        <CheckBoxPreference
            android:persistent="true"
            android:summary="It\'s a checkbox"
            android:title="CheckBoxPreference" />
        <SwitchPreferenceCompat
            android:persistent="true"
            android:summary="It\'s a switch"
            android:title="SwitchPreference" />
        <SeekBarPreference
            android:persistent="true"
            android:summary="It\'s a seek bar"
            android:title="SeekBarPreference" />
        <SeekBarPreference
            android:persistent="true"
            android:summary="It\'s a seek bar without value"
            android:title="SeekBarPreference"
            app:showSeekBarValue="false" />
    </PreferenceCategory>

    <PreferenceCategory android:title="PreferenceCategory using preference_fallback_accent_color below API 21">

    </PreferenceCategory>
    <PreferenceCategory android:title="Another category">
        <EditTextPreference
            android:dialogMessage="This is the dialog\'s message"
            android:inputType="phone"
            android:key="edit_text_test"
            android:persistent="true"
            android:summary="It should be a phone input (it is now!)"
            android:title="EditTextPreference" />
        <ListPreference
            android:dialogTitle="ListPreference dialog"
            android:entries="@array/pref_list_entries"
            android:entryValues="@array/pref_list_values"
            android:key="pref_syncConnectionType"
            android:persistent="true"
            android:summary="%s"
            android:title="ListPreference" />
        <DropDownPreference
            android:dialogTitle="DropDownPreference dialog"
            android:entries="@array/pref_list_entries"
            android:entryValues="@array/pref_list_values"
            android:persistent="true"
            android:summary="%s"
            android:title="DropDownPreference" />
        <PreferenceScreen
            android:key="preference_screen_test"
            android:persistent="true"
            android:title="PreferenceScreen test">
            <Preference
                android:persistent="true"
                android:title="A preference" />
            <SwitchPreferenceCompat
                android:persistent="true"
                android:summary="But this is a SwitchPreferenceCompat"
                android:title="Another one" />
        </PreferenceScreen>
    </PreferenceCategory>
    <PreferenceCategory
        android:key="pref_extras"
        android:title="Extra preferences (available separately)">
        <RingtonePreference
            android:dialogTitle="@string/ringtone_picker_title"
            android:key="pref_ringtone"
            android:persistent="true"
            android:ringtoneType="ringtone"
            android:showDefault="true"
            android:showSilent="true"
            android:summary="Select a nice ringtone"
            android:title="RingtonePreference"
            app:showAdd="true"
            app:summaryHasRingtone="The best ringtone: %s" />
        <TimePickerPreference
            android:key="pref_time"
            android:persistent="true"
            android:summary="Select the perfect time"
            android:title="TimePickerPreference"
            app:hourFormat="auto"
            app:pickerTime="19:35"
            app:summaryHasTime="Selected time: %s" />
        <DatePickerPreference
            android:key="pref_date"
            android:persistent="true"
            android:summary="Select a date"
            android:title="DatePickerPreference"
            app:minDate="08/31/2017"
            app:summaryHasDate="Selected date: %s" />
        <ColorPickerPreference
            android:dialogTitle="@string/color_picker_default_title"
            android:key="pref_color"
            android:persistent="true"
            android:summary="Very nice color"
            android:title="ColorPickerPreference"
            app:size="small" />
    </PreferenceCategory>
    <PreferenceCategory
        android:key="pref_categ"
        android:title="Dynamic category">
        <Preference
            android:key="pref_add"
            android:persistent="true"
            android:summary="Add preference"
            android:title="Preference" />
    </PreferenceCategory>
</PreferenceScreen>

**Comment by Gericop** [comment]
I don't see the `CheckBoxPreference` in your code, but I see that, for example, the `SwitchPreferenceCompat` has no `key` attribute. This means that the preference cannot be saved as there is no key associated with it, which means that when you rotate the phone, there's no persisted value that could be loaded.

**Comment by TCQQ** [comment]
I have not tried before, can you post submit an example, okay? thank you very much.

**Comment by TCQQ** [comment]
Does this library support dynamic addition of data? Because the data returned by the server is dynamic

**Comment by Gericop** [comment]
Here's an example:

```xml
<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">
    <CheckBoxPreference
        android:key="test_checkbox_01"
        android:summary="This will be saved properly"
        android:title="CheckBoxPreference 1" />
    <CheckBoxPreference
        android:key="test_checkbox_02"
        android:persistent="false"
        android:summary="This will not be persisted (persistent = false), but survives rotation (has key)"
        android:title="CheckBoxPreference 2" />
    <CheckBoxPreference
        android:summary="This will not be saved, and does not survive rotation (missing key)"
        android:title="CheckBoxPreference 3" />
</PreferenceScreen>
```

As you can see, **CheckBoxPreference 1** has a `key` attribute set to "test_checkbox_01" and the `persistent` is not set to anything (it's default value is `true`), so the value of the checkbox will be saved to the `SharedPreferences`.

**CheckBoxPreference 2** has a `key` attribute ("test_checkbox_02") but its `persistent` attribute is set to `false`, so it will survive the configuration change (rotation) of the device, but it will not be saved to the `SharedPreferences`. If you leave the app (press the back button) and open it again, the checkbox won't be checked because of this.

**CheckBoxPreference 3** is missing the `key` attribute entirely, so it won't be saved to the `SharedPreferences` nor will it be restored when you rotate the device because the system cannot save the value of it to an instance state due to the missing `key`.

---
As for the other question: what do you mean by dynamic addition of data? The preferences can be added dynamically to the preference screen, the sample app has an example of this in the "Dynamic category" where if you press the Preference (_Add preference_), it will add a new preference to that category.

**Comment by TCQQ** [comment]
The server returns json, and the Preference is displayed according to the json data

**Comment by TCQQ** [comment]
Now "SeekBarPreference" the "seekBarValue" control is displayed on the right side of the progress bar, allowing "seekBarValue" to appear above the progress bar, can you?
![qq 20170922094729](https://user-images.githubusercontent.com/10644057/30726040-130a2960-9f7b-11e7-981f-c1a7d2511d3e.png)

**Comment by Gericop** [comment]
Check out the [`testDynamicPrefs()`](https://github.com/Gericop/Android-Support-Preference-V7-Fix/blob/95f19ae8dd240e00a6d02573f7462d63c80fbac7/app/src/main/java/com/takisoft/preferencefix/MyPreferenceFragment.java#L42) method that shows you how to add preferences dynamically to a category. It's the same when you add the preference to the preference screen which you can get by calling [`getPreferenceScreen()`](https://developer.android.com/reference/android/support/v7/preference/PreferenceFragmentCompat.html#getPreferenceScreen()). If you build the entire preference screen dynamically, just create an almost empty XML with only the `PreferenceScreen` tag in it that will be your root preference. Load it like you do now, then use `getPreferenceScreen()` to get a reference to it, then just add the new preferences by calling [`addPreference(...)`](https://developer.android.com/reference/android/support/v7/preference/PreferenceGroup.html#addPreference(android.support.v7.preference.Preference)).

**Comment by TCQQ** [comment]
Ok thank you,I check now

**Comment by TCQQ** [comment]
Now "SeekBarPreference" the "seekBarValue" control is displayed on the right side of the progress bar, allowing "seekBarValue" to appear above the progress bar, can you?
![qq 20170922094729](https://user-images.githubusercontent.com/10644057/30726214-148abccc-9f7c-11e7-82c2-6af0999fe092.png)

**Comment by Gericop** [comment]
No, you need a custom layout XML to allow the value holder `TextView` to be moved to a different place. It'd be best to copy and modify the `preference_widget_seekbar_material.xml` file(s).

> Note that those files are probably not needed anymore as it seems like the Google guys fixed the layout of the seekbar, but I haven't tested it yet.

**Comment by TCQQ** [comment]
ok, i saw it, but "seekBarValue" is not real-time, can  make "seekBarValue" with "SeekBar" drag the number of real-time changes value?

**Comment by Gericop** [comment]
No, that would require some heavy reflection usage throughout the whole `SeekBarPreference` class.

**Comment by TCQQ** [comment]
But adjust the volume is generally real-time

**Comment by Gericop** [comment]
Well, tell this to the Google devs 😄 But IMO, it's not a huge problem if you don't display the exact value because no one cares about it, the bar itself represents the volume and probably all users can understand the meaning of it without seeing the actual value.

**Comment by TCQQ** [comment]
I changed the layout (layout-v21, layout-v14) layout, deleted (layout-v17), I think now looks much better than before
![ibin f7 1 nvhp_8g4 ao3](https://user-images.githubusercontent.com/10644057/30727273-dd852b7a-9f82-11e7-902b-ff3858424ba1.png)

[layout.zip](https://github.com/Gericop/Android-Support-Preference-V7-Fix/files/1323302/layout.zip)

**Comment by Gericop** [comment]
I won't change it since no one has had problems with it so far and anyone can easily override the layout files (just like you did) if they have special needs.

**Comment by TCQQ** [comment]
Yes, honestly now the layout looks better than before, so do not need to add the code to the project, directly use the gradle link can be used directly, more convenient

**Comment by Gericop** [comment]
By the way, the no real-time-update probably has practical reasons: if it updated the value real-time, the `SharedPreferences` would get a ton of updates, stressing the device unnecessarily. Updating only the `TextView` would do no good probably since, as you said, you wanted to listen to the changes which only happen before the value gets persisted. If the user moved the thumb from 100 to 0, it would send 100 change notifications, again, stressing the device. It might even make the UI become laggy.

Of course, it would be possible to update the value on the UI and create a separate interface for listening for updates without persisting the value, but that's really some serious effort just to display and use the value in real-time, yet it is probably not even needed most of the time.

**Comment by TCQQ** [comment]
Now there is a library that implements real-time updates and you can look at his code

https://github.com/consp1racy/android-support-preference#dividers

**Comment by Gericop** [comment]
I know how to do it, the only problem with the `SeekBarPreference` is that everything is private so simply extending the class and overriding a few things is not possible. Instead, one would have to basically copy the whole class then make the necessary changes. This makes the updating process time consuming, since the new class should reflect every change from the official class, which is kind of a diffing between two classes.

As you can see, the creator of that lib also recreated every single preference type which means a bigger library and slower updates. That library uses 23.x.x dependencies (according to [this](http://www.methodscount.com/?lib=net.xpece.android%3Asupport-preference%3A1.2.8) page) while this one is and always has been on the latest version (26.1.0 at the time of writing this) within days (or sometimes just hours) since there's usually no need to rewrite complex classes to make it work when a new support lib version is released (which happens usually once or twice a month; major releases happen yearly).

**TL;DR:** Re-implementing every preference to make them fit the custom needs results in bigger lib size and slower updates.

**Comment by TCQQ** [comment]
I think you're right, you thinking is very comprehensive

You have time can look at i in above send the layout, I tested

**Comment by TCQQ** [comment]
If use svn as an icon, can dynamically change icon color in the code?

Like this：https://github.com/AAkira/CompoundIconTextView

**Comment by Gericop** [comment]
You cannot use SVGs in Android but vector drawables (they are not the same), just to be clear.

However, the official `Preference` implementation does not work with vector drawables on devices with API version lower than 21 (so pre-lollipop) because it uses `ContextCompat` for inflating drawables and that's basically just a wrapper around getting them on different API levels. If you want to use vector drawables as icons, you have a few choices.

**Method 1**
Use [`AppCompatResources.getDrawable(...)`](https://developer.android.com/reference/android/support/v7/content/res/AppCompatResources.html#getDrawable(android.content.Context,%20int)) to inflate the icons and set the returned drawables as the appropriate preferences' icons using `setIcon(...)`. The drawback is that you need to setup all of the preferences programmatically, which can be cumbersome.

**Method 2**
Use [`VectorDrawableCompat.create(...)`](https://developer.android.com/reference/android/support/graphics/drawable/VectorDrawableCompat.html#create(android.content.res.Resources,%20int,%20android.content.res.Resources.Theme)) to create the vector drawable. The difference between this and _method 1_ is that you cannot use this to inflate regular (i.e. non-vector) drawables. If you mix them, use _method 1_ instead.

**Method 3**
Use [`AppCompatDelegate.setCompatVectorFromResourcesEnabled(...)`](https://developer.android.com/reference/android/support/v7/app/AppCompatDelegate.html#setCompatVectorFromResourcesEnabled(boolean)) to enable vector drawable inflation automatically for every drawable container.

```java
static {
    AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
}
```

Note, however, that this solution can introduce memory leaks and other problems to your app as the documentations states it:
> This feature defaults to disabled, since enabling it can cause issues with memory usage, and problems updating Configuration instances. If you update the configuration manually, then you probably do not want to enable this. You have been warned.

**Comment by Gericop** [comment]
I opened [an issue](https://issuetracker.google.com/issues/66695659) to try to convince the devs to change `ContextCompat` to `AppCompatResources` in order to enable vector drawable loading with only the `vectorDrawables.useSupportLibrary = true` present in the gradle build file. If you want to, you can _star_ it to show interest in the problem.

**Comment by TCQQ** [comment]
Ok, thank you i will to try

**Comment by TCQQ** [comment]
I'm not add this library：compile 'com.takisoft.fix:preference-v7-colorpicker:26.1.0.0'，but app error，what reason?

Code：

<?xml version="1.0" encoding="utf-8"?>
<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto">

    <PreferenceCategory android:title="First category">

        <ColorPickerPreference
            android:key="pref_color"
            android:persistent="false"
            android:summary="Very nice color" />

    </PreferenceCategory>

</PreferenceScreen>

Error Log：

java.lang.NullPointerException: Attempt to invoke virtual method 'float android.view.View.getY()' on a null object reference
                                                                   at com.takisoft.fix.support.v7.preference.PreferenceFragmentCompatDividers$DividerItemDecoration.onDrawOver(PreferenceFragmentCompatDividers.java:288)
                                                                   at android.support.v7.widget.RecyclerView.draw(RecyclerView.java:3991)
                                                                   at android.view.View.updateDisplayListIfDirty(View.java:18069)
                                                                   at android.view.View.draw(View.java:18847)
                                                                   at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                   at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                   at android.view.View.updateDisplayListIfDirty(View.java:18060)

**Comment by TCQQ** [comment]
I would like to use the official "ColorPickerPreference", what should I do?

**Comment by TCQQ** [comment]
Feedback bug：
EditTextPreference the dialog click “ok” or “cancel” after, the keyboard can't close

**Comment by TCQQ** [comment]
I only add preference-v7 library in my project, When i click in "EditTextPreference" the dialog  "ok" or "cencel" button after,  the keyboard cant close, i update EditTextPreferenceDialogFragmentCompat after, this problem will be repair

class：EditTextPreferenceDialogFragmentCompat
 
   public void onDialogClosed(boolean positiveResult) {
        if (positiveResult) {
            String value = this.mEditText.getText().toString();
            if (this.getEditTextPreference().callChangeListener(value)) {
                this.getEditTextPreference().setText(value);
            }
        }
        toggleSoftInput(getContext());
    }

    public static void toggleSoftInput(Context context) {
        InputMethodManager imm = (InputMethodManager) context.getApplicationContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm == null) return;
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
    }

**Comment by Gericop** [comment]
What Android version and device is this? Can you reproduce it with the sample app too?

**Comment by TCQQ** [comment]
Pixel XL 8.0

**Comment by Gericop** [comment]
Can you reproduce it with the sample app too?

**Comment by TCQQ** [comment]
Wait

**Comment by TCQQ** [comment]
Done

**Comment by TCQQ** [comment]
Ready

**Comment by TCQQ** [comment]
check
[Uploading MyApplication.zip…]()

**Comment by Gericop** [comment]
I'm asking if you could reproduce it on your device using the sample app. Clone the project and install the sample app and see if you see the problem happening there. If not, there's probably something wrong with your app, otherwise there's a problem with the lib. I cannot reproduce it using the sample app, it works both on my device (7.0) and in the emulator (8.0).

**Comment by TCQQ** [comment]
Yes i tested

**Comment by TCQQ** [comment]
Please wait, uploading

**Comment by Gericop** [comment]
Why do you need to upload something for this? Just tell me if the exception was thrown or not.

**Comment by TCQQ** [comment]
[MyApplication.zip](https://github.com/Gericop/Android-Support-Preference-V7-Fix/files/1326908/MyApplication.zip)

**Comment by TCQQ** [comment]
Yes

**Comment by TCQQ** [comment]
This is error sample app

**Comment by TCQQ** [comment]
java.lang.NullPointerException: Attempt to invoke virtual method 'float android.view.View.getY()' on a null object reference
                                                                                at com.takisoft.fix.support.v7.preference.PreferenceFragmentCompatDividers$DividerItemDecoration.onDrawOver(PreferenceFragmentCompatDividers.java:288)
                                                                                at android.support.v7.widget.RecyclerView.draw(RecyclerView.java:3991)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18069)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                                at android.view.View.draw(View.java:18847)
                                                                                at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                                at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                                at android.view.View.draw(View.java:19122)
                                                                                at com.android.internal.policy.DecorView.draw(DecorView.java:785)
                                                                                at android.view.View.updateDisplayListIfDirty(View.java:18069)
                                                                                at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:643)
                                                                                at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:649)
                                                                                at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:757)
                                                                                at android.view.ViewRootImpl.draw(ViewRootImpl.java:2980)
                                                                                at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:2794)
                                                                                at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2347)
                                                                                at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1386)
                                                                                at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6733)
                                                                                at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
                                                                                at android.view.Choreographer.doCallbacks(Choreographer.java:723)
                                                                                at android.view.Choreographer.doFrame(Choreographer.java:658)
                                                                                at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
                                                                                at android.os.Handler.handleCallback(Handler.java:789)
                                                                                at android.os.Handler.dispatchMessage(Handler.java:98)
                                                                                at android.os.Looper.loop(Looper.java:164)
                                                                                at android.app.ActivityThread.main(ActivityThread.java:6541)
                                                                                at java.lang.reflect.Method.invoke(Native Method)
                                                                                at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
                                                                                at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)

**Comment by TCQQ** [comment]
this is all error log

**Comment by Gericop** [comment]
Please don't post single word / very short messages. They all belong together, and this is not a chat.

**Comment by TCQQ** [comment]
Ok

**Comment by Gericop** [comment]
I fixed the bug, it will be released soon.

**Comment by TCQQ** [comment]
Awesome

**Comment by TCQQ** [comment]
I wait you
![8oevovti oo0cvl 4m6ex8](https://user-images.githubusercontent.com/10644057/30777979-b4f48844-a0fa-11e7-851a-0784ef02ac57.png)

**Comment by Gericop** [comment]
Released v**26.1.0.1**:
```gradle
compile 'com.takisoft.fix:preference-v7:26.1.0.1'
```

Watch out that `ColorPickerPreference` now defaults to auto column number instead of 3. This usually looks good, but the flexbox layout has an issue on API 16 and earlier, so the last items in the rows look overlapped by the dialog's side padding. It will be fixed as soon as the flexbox layout v0.3.1 version gets released.

**Comment by TCQQ** [comment]
Ok，i test now

**Comment by Gericop** [comment]
Also, if you don't set a category's title, it will be hidden, but the dividers's will be drawn between the categories (if you set the appropriate divider flags). Check out the release notes for more info.

**Comment by TCQQ** [comment]
Ok wait

**Comment by TCQQ** [comment]
Error：

 Process: com.changrui.myapplication, PID: 7437
                                                                          java.lang.NullPointerException: Attempt to invoke virtual method 'int android.support.v7.widget.RecyclerView$Adapter.getItemCount()' on a null object reference
                                                                              at com.takisoft.fix.support.v7.preference.PreferenceFragmentCompatDividers$DividerItemDecoration.onDrawOver(PreferenceFragmentCompatDividers.java:264)
                                                                              at android.support.v7.widget.RecyclerView.draw(RecyclerView.java:3991)
                                                                              at android.view.View.updateDisplayListIfDirty(View.java:18069)
                                                                              at android.view.View.draw(View.java:18847)
                                                                              at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                              at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                              at android.view.View.updateDisplayListIfDirty(View.java:18060)
                                                                              at android.view.View.draw(View.java:18847)
                                                                              at android.view.ViewGroup.drawChild(ViewGroup.java:4214)
                                                                              at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4000)
                                                                              at android.view.View.updateDisplayListIfDirty(View.java:18060)

**Comment by Gericop** [comment]
Seriously, I just tested it with both your project and the sample app on API 26, API 17 and API 15, and all of them worked fine. Is this still the example you sent me or your _real_ project?

**Comment by TCQQ** [comment]
![image](https://user-images.githubusercontent.com/10644057/30778436-84ab8aee-a108-11e7-8952-914d303bd6a7.png)

**Comment by TCQQ** [comment]
Yes, I only update "com.takisoft.fix:preference-v7" version number

**Comment by Gericop** [comment]
Test **26.1.0.2**.

**Comment by TCQQ** [comment]
wait

**Comment by TCQQ** [comment]
![image](https://user-images.githubusercontent.com/10644057/30778486-d03710cc-a109-11e7-9716-985ddde59bc2.png)
No error，but layout is null

**Comment by TCQQ** [comment]
Cannot display layout

**Comment by Gericop** [comment]
Do you use `ColorPickerPreference` or any of the extras?

**Comment by TCQQ** [comment]
Yes, but i was not update other code, this is xml code:

<?xml version="1.0" encoding="utf-8"?>
<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android">

    
    <PreferenceCategory android:title="First category">
        <Preference
            android:persistent="false"
            android:summary="This is just a normal preference"
            android:title="Preference" />
        <Preference
            android:enabled="false"
            android:persistent="false"
            android:summary="This is just a disabled preference"
            android:title="Preference 2" />
        <EditTextPreference
            android:dialogMessage="This is the dialog\'s message"
            android:inputType="phone"
            android:key="edit_text_test"
            android:persistent="false"
            android:summary="It should be a phone input (it is now!)"
            android:title="EditTextPreference" />
    </PreferenceCategory>


    
    <PreferenceCategory
        android:key="pref_extras"
        android:title="Extra preferences (available separately)">
        <ColorPickerPreference
            android:dialogTitle="Title"
            android:key="pref_color"
            android:persistent="false"
            android:summary="Very nice color"
            android:title="ColorPickerPreference" />
    </PreferenceCategory>

</PreferenceScreen>

**Comment by TCQQ** [comment]
If i annotations “ColorPickerPreference” layout will show

![image](https://user-images.githubusercontent.com/10644057/30778508-7fead33c-a10a-11e7-88c0-7b53fa5b26d9.png)

**Comment by Gericop** [comment]
Well, the problem is easy to fix. First of all, set your logcat filtering to some lower level, debug for example, so you'd see that `ColorPickerPreference` couldn't be inflated because it was not found. 

The more important thing, though, is to actually **add the dependency to the project**. The README.md file clearly says:

> There are additional preferences not part of the official support library, but decided to add them to some extra libraries. You can add all of them to your project using `compile 'com.takisoft.fix:preference-v7-extras:26.1.0.2'` or one or more groups:
> [...]
> `compile 'com.takisoft.fix:preference-v7-colorpicker:26.1.0.2'`

So you're missing this dependency, thus your XML cannot be inflated.

**Comment by TCQQ** [comment]
But now the official colorpicker can not be used

**Comment by Gericop** [comment]
What do you mean? There's no official `ColorPickerPreference`. The preference manager has been modified in a way that it recognizes custom preferences without their fully qualified names (i.e. without the package name).

**Comment by TCQQ** [comment]
I tested (RingtonePreference, TimePickerPreference, DatePickerPreference, ColorPickerPreference) have the same problem

**Comment by TCQQ** [comment]
The question now is google provided controls are not available

**Comment by Gericop** [comment]
Just read the README (hence the name: _read me_), please. You have to add the appropriate dependencies.

### Example 1 - You want to use all of the custom preferences
```gradle
compile 'com.takisoft.fix:preference-v7:26.1.0.2' // the base artifact
compile 'com.takisoft.fix:preference-v7-extras:26.1.0.2' // + every custom preference's artifact
```
---


### Example 2 - You want to use only the `ColorPickerPreference`
```gradle
compile 'com.takisoft.fix:preference-v7:26.1.0.2' // the base artifact
compile 'com.takisoft.fix:preference-v7-colorpicker:26.1.0.2' // + the colorpicker's artifact
```

---

If you want to use the _extra_ preferences, you need to add them to your dependencies as _extra_ dependencies, not replacing the base project (`preference-v7`).

**Comment by TCQQ** [comment]
“com.takisoft.fix:preference-v7-extras:26.1.0.2” cant add. If add “com.takisoft.fix:preference-v7-extras:26.1.0.2” can use google provided controls？

![image](https://user-images.githubusercontent.com/10644057/30778662-32c3f90e-a10e-11e7-913a-52627d1da0f7.png)

**Comment by Gericop** [comment]
Open the terminal in Android Studio and type in this:

`gradlew --refresh-dependencies`

**Comment by TCQQ** [comment]
ok

**Comment by TCQQ** [comment]
I add “com.takisoft.fix:preference-v7-extras:26.1.0.2” after, I can use google provided controls, right？

**Comment by Gericop** [comment]
You need both `preference-v7` and `preference-v7-extras`, and yes, the official ones will be available after that, too. The `preference-v7-extras` contains (for the time being) the `RingtonePreference`, `DatePickerPreference`, `TimePickerPreference`, and `ColorPickerPreference`.

**Comment by TCQQ** [comment]
![image](https://user-images.githubusercontent.com/10644057/30778885-14d7e4e6-a113-11e7-9428-6f53dab8befe.png)
Now AndroidStudio in the implementation of the command you just sent me, I can now delete the red arrow library?

**Comment by Gericop** [comment]
If you want to use `ColorPickerPreference`, then you need it (or the `preference-v7-colorpicker` artifact).

**Comment by TCQQ** [comment]
I executed "gradlew --refresh-dependencies", but it is still cant to use google the "ColorPickerPreference" now

**Comment by Gericop** [comment]
What's in your dependencies now?

**Comment by TCQQ** [comment]
Can use the controls provided by Google?

**Comment by TCQQ** [comment]
![image](https://user-images.githubusercontent.com/10644057/30779063-d6b0ac7a-a117-11e7-90e7-1510a47112d4.png)

**Comment by Gericop** [comment]
I told you that you need BOTH dependencies:

```gradle
compile 'com.takisoft.fix:preference-v7:26.1.0.2' // the base artifact with the base preferences
compile 'com.takisoft.fix:preference-v7-colorpicker:26.1.0.2' // + the ColorPickerPreference
```

**Comment by TCQQ** [comment]
I know you mean, but if i want use android comes with "colorpicker", how should i add?

**Comment by Gericop** [comment]
The Android official (or support, doesn't matter) preferences does not have color picker. If you mean a separate color picker widget, it won't interfere with it.

**Comment by TCQQ** [comment]
![image](https://user-images.githubusercontent.com/10644057/30779157-3b6389d2-a11b-11e7-929f-d99a85fc9566.png)
![image](https://user-images.githubusercontent.com/10644057/30779158-44073cbe-a11b-11e7-902b-1ca1a97b1af8.png)

But now the "colorpicker" and Material are inconsistent
https://materialdoc.com/components/pickers/#color-picker

**Comment by Gericop** [comment]
No, it fits the available dialog space. If you want to force it to have 5 columns, you can do it by setting `app:columns="5"` on the `ColorPickerPreference`, but it will probably result in a bad (cut-off) layout. The [wiki article](https://github.com/Gericop/Android-Support-Preference-V7-Fix/wiki/Preference-types#colorpickerpreference) says it all.

Also, there's **no official color picker widget in Android**, so saying that it is _inconsistent_ does not make any sense (_[materialdoc.com](https://materialdoc.com) is not an official Google site_). The color picker widget is an open source, _separate_ project available from the [Google repo](https://android.googlesource.com/platform/frameworks/opt/colorpicker/). My implementation uses it with a slight upgrade: instead of a "fixed number of columns" layout it employs the [FlexboxLayout](https://github.com/google/flexbox-layout) so that it can automatically fit the available space. It will probably fit more once the new version of the flexbox gets released as it fixes a couple things and I've also changed a few things in the dialog that makes it really stretch the dialog to a bigger size, but I don't want to release it until the flexbox gets fixed.

**Comment by Gericop** [comment]
I'm closing this issue. If you find other bugs / have issues, just open a new one.

**Comment by TCQQ** [comment]
Okay, thankyou
